docs/common-sdk-design.md

```diff
@@ -0,0 +1,351 @@
+# Common SDK infrastructure dev design
+The purpose of this doc is to serve as a guidance for software engineers to implement a solution defined below as "the common SDK infrastructure". It contains basic analysis of the problem, proposal for improvements, goals, requirements, architecture and a list of technical challenges with suggestions for implementation.
+
+## Problem analysis
+There is nothing fundamentally wrong with current PlayFab SDKs and the SDK generation process. They are technically designed and implemented well, relatively simple, efficient and address most of our today's customers' needs. The simplicity of current implementation - while being a powerful benefit - also has a cost. There are certain limitations that may be standing in a way of even greater adoption of the platform by customers with specific needs or extending business opportunities provided by the platform:
+* Monolithic, one-for-all formation of current SDKs results in:
+    * Limited configurability (ability to select which components are needed or should be excluded from the SDK)
+    * Limited customization options (ability to easily replace standard components with custom ones in the SDK)
+    * Limited extensibility (ability to easily add new external functionality to the SDK)
+* No possibility to build a custom SDK
+* No easy and configurable way to:
+    * Use an alternative service platform (e.g. a different data ingestion system)
+    * Use a different network transport/protocol (unless already built in the SDK)
+    * Use different elements of data processing stack on the client side (e.g. specific data formatting, serialization, compression, encryption)
+    * Use alternative client-side API
```

Comment:
How is this different from what you described under "Monolithic..." ?

Reply:
I put it in a separate bullet point because it is one of our major goals down the road, so it focuses attention on a feature that is currently not supported but can be with a modular SDK

---
In reply to: [194559323](https://github.com/PlayFab/SDKGenerator/pull/134#discussion_r194559323) [](ancestors = 194559323)